### PR TITLE
Port IconInfo to Vala

### PR DESCRIPTION
**Pull request by tintou**

### CODE REVIEW

File: libcore/IconInfo.vala [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    private struct LoadableIconKey {
+        GLib.Icon icon;
+        int size;
+    }
+
+    private struct ThemedIconKey {
+        string filename;
+        int size;
+    }
+
```
Comment (jeremypw) [inline libcore/IconInfo.vala]: Could make the keys simple classes to avoid having to box them.

File: libcore/IconInfo.vala [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    private static uint loadable_icon_key_hash (LoadableIconKey? key) {
+        return (key.icon.hash () ^ key.size);
+    }
+
+    private static bool loadable_icon_key_equal (LoadableIconKey? a, LoadableIconKey? b) {
+        return (a.size == b.size && a.icon.equal (b.icon));
+    }
+
+    private static uint themed_icon_key_hash (ThemedIconKey? key) {
+        return (key.filename.hash () ^ key.size);
```
Comment (jeremypw) [inline libcore/IconInfo.vala]: Appears to dereference a potential null

File: libcore/IconInfo.vala [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        if (filename != null) {
+            filename = GLib.Path.get_basename (filename);
+            var last = filename.last_index_of_char ('.');
+            icon_name = filename.substring (0, last);
+        }
+    }
+
+    construct {
+        last_use_time = GLib.get_monotonic_time ();
+        schedule_reap_cache ();
```
Comment (jeremypw) [inline libcore/IconInfo.vala]: Not sure why?

File: libcore/IconInfo.vala [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+                    critical (e.message);
+                }
+            }
+
+            if (pixbuf != null) {
+                icon_info = new IconInfo.for_pixbuf (pixbuf);
+                loadable_icon_cache.insert (loadable_key, icon_info);
+            }
+
+            return null;
```
Comment (jeremypw) [inline libcore/IconInfo.vala]: null here differs from .c file, which returns icon_info.

File: libcore/IconInfo.vala (hunk 3 above)
Comment (jeremypw) [reply libcore/IconInfo.vala]: On reflection, this replaces the reaping that in the c version was triggered by pixbuf_toggle_notify (i.e when a pixbuf became unused for a certain time).

**Comment by jeremypw** [comment]
The new iconinfo test would have picked up the problem with the null return from iconinfo.lookup ().

**Comment by jeremypw** [comment]
Not sure why Travis CI is failing - running ctest locally passes OK.  Will investigate

**Comment by tintou** [comment]
@jeremypw Alright, seems like this is good this time, looks like it's ready to merge IMO

**Comment by jeremypw** [comment]
OK lets go for it.

**Pull request merged**